File: DataAnalysis/ChessBrawl.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace DataAnalysis
{
	public class Node
	{
		public string Name { get; set; }
		public int Turn { get; }
		public List<Node> Children = new List<Node>();
		public int Value { get; set; }
		public int Wins { get; set; }
		public int Losses { get; set; }

		public Node(string name, int turn)
		{
			Name = name;
			Turn = turn;
		}

		public Node GetChild(string name, int turn)
		{
			var child = Children.FirstOrDefault(x => x.Name == name && x.Turn == turn);
			if(child == null)
			{
				child = new Node(name, turn);
				Children.Add(child);
			}
			return child;
		}
	}

	internal class ChessBrawl
	{
		public int TotalGames { get; set; }
		public int WhiteWins { get; set; }
		public int[] CoinWins = new int[100];
		public int[] CoinLosses = new int[100];
		private List<Game> _games;
		private readonly string _inputFile;
		private readonly string _outputFile;

		public ChessBrawl(string inputFile, string outputFile)
		{
			_inputFile = inputFile;
			_outputFile = outputFile;
		}

		public void CreateTree(int depth)
		{
			_games = ReadGames(_inputFile).ToList();
			var root = new Node("Root", 0);
			foreach(var game in _games)
			{
				var currNode = root;
				for(var turn = 1; turn <= game.Turns; turn++)
				{
					var player = turn%2 == 0 ? game.P2 : game.P1;
					var plays = player.Plays.Where(x => x.Turn == turn).ToList();
					var name = plays.Count == 0 ? "PASS" : string.Join(",", plays.Select(x => x.CardId));
					currNode = currNode.GetChild(name, turn);
					currNode.Value++;
					if(player.Result == "WON")
						currNode.Wins++;
					else currNode.Losses++;
				}
			}
			using(var sw = new StreamWriter(_outputFile))
				foreach(var line in PrintNode(root, depth))
					sw.WriteLine(line);
		}

		private IEnumerable<string> PrintNode(Node node, int depthThreshold)
		{
			yield return $"T{node.Turn} {node.Name} {node.Value} {Math.Round(100.0 * node.Wins / (node.Wins + node.Losses), 2)}%";
			foreach(var child in node.Children)
			{
				if(child.Turn > depthThreshold) continue;
				foreach(var val in PrintNode(child, depthThreshold))
					yield return val;
			}
		}

		public void Foo()
		{
			_games = ReadGames(_inputFile).ToList();
			foreach(var game in _games)
			{
				TotalGames++;
				var whiteWon = false;
				if(game.P1.Result == "WON")
				{
					WhiteWins++;
					whiteWon = true;
				}
				var coinTurn = game.P2.Plays.FirstOrDefault(x => x.CardId == "GAME_005")?.Turn ?? 99;
				(whiteWon ? CoinLosses : CoinWins)[coinTurn]++;

			}
			Console.WriteLine("Total Games: " + TotalGames);
			Console.WriteLine($"White Wins: {WhiteWins} ({100.0*WhiteWins/TotalGames}%)");
			for(int turn = 2; turn < 100; turn += 2)
			{
				var wins = CoinWins[turn];
				var losses = CoinLosses[turn];
				if(wins + losses == 0)
					break;
				Console.WriteLine($"Turn {turn/2} Coin winrate: {Math.Round(100.0*wins/(wins + losses), 2)}% ({wins + losses} games)");
			}
			var noCoinWins = CoinWins[99];
			var noCoinLosses = CoinLosses[99];
			Console.WriteLine($"No Coin played winrate: {Math.Round(100.0*noCoinWins/(noCoinWins + noCoinLosses), 2)}% ({noCoinWins + noCoinLosses} games)");
			Console.WriteLine(string.Join(", ", CoinWins.Take(10)));
			Console.WriteLine(string.Join(", ", CoinLosses.Take(10)));
		}

		private IEnumerable<Game> ReadGames(string inputFile)
		{
			var invalidGames = 0;
			using(var reader = new StreamReader(inputFile))
			{
				var game = new Game();
				while(!reader.EndOfStream)
				{
					var line = reader.ReadLine();
					if(game.Read(line))
					{
						if(game.P1.Hero.StartsWith("KAR_"))
							yield return game;
						else invalidGames++;
						game = new Game();
					}
				}
			}
			Console.WriteLine("Invalid Games: " + invalidGames);
		}
	}

	internal class Game
	{
		private int _index;
		public string Id { get; set; }
		public int FirstPlayer;
		public int FriendlyPlayer;
		public int Turns;
		public Player P1 { get; set; }
		public Player P2 { get; set; }

		public bool Read(string line)
		{
			var values = line.Split(',');
			if(_index == 0)
			{
				Id = values[0];
				int.TryParse(values[2], out FirstPlayer);
				int.TryParse(values[3], out FriendlyPlayer);
				int.TryParse(values[4], out Turns);
			}
			else if(_index == 1)
				P1 = Player.Read(line);
			else if(_index == 2)
			{
				P2 = Player.Read(line);
				return true;
			}
			_index++;
			return false;
		}
	}

	public class Player
	{
		public string Hero { get; set; }
		public string Result { get; set; }
		public int Fatigue;
		public int HeroPowerUses;
		public IEnumerable<Action> Draws { get; set; }
		public IEnumerable<Action> Plays { get; set; }

		public static Player Read(string line)
		{
			var values = line.Split(',');
			var player = new Player
			{
				Hero = values[2],
				Result = values[3],
				Draws = GetActions(values[6]),
				Plays = GetActions(values[7]),
			};
			int.TryParse(values[4], out player.Fatigue);
			int.TryParse(values[5], out player.HeroPowerUses);
			return player;
		}

		private static IEnumerable<Action> GetActions(string line) => line.Split('|').Select(Action.Parse);
	}

	public class Action
	{
		public string CardId;
		public int Turn;

		public static Action Parse(string actionString)
		{
			var action = new Action();
			var parts = actionString.Split(':');
			action.CardId = parts[0];
			if(parts.Length > 1)
				int.TryParse(parts[1], out action.Turn);
			return action;
		}
	}
}

File: FDataAnalaysis/ChessBrawl.fs
module ChessBrawl

let GetInt (s:string[], i:int) = 
    if s.Length <= i then None 
    else match System.Int32.TryParse(s.[i]) with 
            | (true, int) -> Some(int) 
            | _ -> None

type Action (str:string) =
    let parts = str.Split[|':'|]
    member this.CardId = parts.[0]
    member this.Turn = (parts, 1) |> GetInt 

type Player (str:string) =
    let parts = str.Split[|','|]
    let GetActions (s:string) = s.Split[|'|'|] |> Seq.map(Action)
    member this.Hero = parts.[2]
    member this.Result = parts.[3]
    member this.Draws = GetActions parts.[6]
    member this.Plays = GetActions parts.[7]

type Game (str:string list) =
    let meta = str.[2].Split[|','|]
    member this.Id = meta.[0]
    member this.FirstPlayer = (meta, 2) |> GetInt
    member this.FriendlyPlayer = (meta, 3) |> GetInt
    member this.Turns = (meta, 4) |> GetInt
    member this.P1 = Player(str.[1])
    member this.P2 = Player(str.[0])

type Node (name:string, turn:int) =
    member this.Name = name
    member this.Turn = turn
    member val Value = 0 with get, set
    member val Wins = 0 with get, set
    member val Losses = 0 with get, set
    member val Children = List.empty<Node> with get, set
    member this.GetChild (n:string, t:int) = 
        let mutable child = this.Children |> List.tryFind(fun x -> x.Name = n && x.Turn = t)
        if child = None then child <- Node(n, t) |> Some; this.Children <- child.Value :: this.Children
        child.Value.Value <- child.Value.Value + 1 
        (child.Value)

let Read (inputFile:string) = 
    seq {
        let mutable lines = List.empty
        use sr = new System.IO.StreamReader(inputFile)
        while not sr.EndOfStream do
            lines <- sr.ReadLine() :: lines
            if lines.Length = 3 then 
                let game = Game lines
                if game.P1.Hero.StartsWith "KAR_" then yield game
                lines <- List.empty
    }

let GenTree (games:Game seq, depth:int) =
    let mutable root = Node("Root", 0)
    for game in games do
        let mutable node = root
        for turn in [1..min game.Turns.Value depth] do
            let player = if turn%2 = 0 then game.P2 else game.P1
            let plays = player.Plays |> Seq.filter(fun x -> x.Turn = Some(turn))
            let name = if Seq.isEmpty(plays) then "None" else plays |> Seq.map(fun x -> x.CardId) |> String.concat ","
            node <- node.GetChild(name, turn)
            if player.Result = "WON" then node.Wins <- node.Wins+1 else node.Losses <- node.Losses + 1
    (root)

let CardWinrateByTurn (games:Game seq, cardIds:string[]) =
    let queenStats = Array.replicate 35 (0, 0)
    let getQueenAction (game:Game, result:string) = 
        match result with 
        | r when r = game.P1.Result -> game.P1.Plays
        | _ -> game.P2.Plays
        |> Seq.tryFind(fun x -> cardIds |> Array.contains x.CardId), result = "WON"        
    let updateStats (action:Action option, won:bool) =
        if action.IsSome then
            let wins, losses = queenStats.[action.Value.Turn.Value]
            let w, l = if won then (1, 0) else (0, 1)
            queenStats.[action.Value.Turn.Value] <- (wins + w, losses + l)
    for game in games do
        (game, "WON") |> getQueenAction |> updateStats |> ignore
        (game, "LOST") |> getQueenAction  |> updateStats |> ignore
    queenStats |> Array.map(fun (wins, losses) -> (float wins / float (wins + losses), wins + losses)) |> Array.indexed |> printfn "%A"
    0

let rec Stringify (n:Node) = 
    seq {
        if n.Wins > 0 || n.Losses > 0 then 
            let winrate = System.Math.Round(float n.Wins / float (n.Wins + n.Losses), 2)
            yield [string n.Turn; n.Name; string n.Value; string winrate] |> String.concat " "
        for child in n.Children do
            for node in Stringify child do yield node
    }
